daemon/cluster/listen_addr.go

```diff
@@ -0,0 +1,159 @@
+package cluster
+
+import (
+	"errors"
+	"fmt"
+	"net"
+)
+
+func resolveListenAddr(specifiedInterface, specifiedAddr string) (string, error) {
+	specifiedHost, specifiedPort, err := net.SplitHostPort(specifiedAddr)
+	if err != nil {
+		return specifiedAddr, nil
+	}
+	ip := net.ParseIP(specifiedHost)
+	if ip != nil && ip.IsUnspecified() {
+		// The user has not provided a listen address.
+		if specifiedInterface != "" {
+			interfaceAddr, err := resolveInterfaceAddr(specifiedInterface)
+			if err != nil {
+				return "", err
+			}
+			return net.JoinHostPort(interfaceAddr.String(), specifiedPort), nil
+		}
+
+		systemAddr, err := resolveSystemAddr()
+		if err != nil {
+			return "", err
+		}
+		return net.JoinHostPort(systemAddr.String(), specifiedPort), nil
+	} else if specifiedInterface != "" {
+		return "", errors.New("cannot specify both --listen-interface and --listen-addr")
+	}
+
+	return specifiedAddr, nil
+}
+
+func resolveInterfaceAddr(specifiedInterface string) (net.IP, error) {
+	// Use a specific interface's IP address.
+	intf, err := net.InterfaceByName(specifiedInterface)
```

Comment:
The current version of the PR allows you to specify `127.0.0.1` or `lo`. I prefer to support it because if you are setting up a swarm that will only have one node, you can use `--advertise-addr 127.0.0.1` or `--advertise-addr lo` to satisfy the condition of having an advertise address. On an embedded system, you might not actually have any other interfaces or addresses.

But I suppose I'm open to blacklisting these if people think it it's better than allowing them.